spiketoolkit/validation/quality_metric_classes/noise_overlap.py

```diff
@@ -23,47 +26,75 @@ def __init__(self, metric_data):
         if not metric_data.has_recording():
             raise ValueError("MetricData object must have a recording")
 
-    def compute_metric(self, max_spikes_per_unit_for_noise_overlap, num_features, num_knn, **kwargs):
+    def compute_metric(self, num_channels_to_compare, max_spikes_per_unit_for_noise_overlap,
+                        num_features, num_knn, **kwargs):
+
+        # update keyword arg in case it's already specified to something
+        kwargs['max_spikes_per_unit'] = max_spikes_per_unit_for_noise_overlap
+        kwargs['recompute_info'] = True
         params_dict = update_all_param_dicts_with_kwargs(kwargs)
         save_property_or_features = params_dict['save_property_or_features']
         seed = params_dict['seed']
 
+        # first, get waveform snippets of every unit (at most n spikes)
+        # waveforms = List (units,) of np.array (n_spikes, n_channels, n_timepoints)
         waveforms = st.postprocessing.get_unit_waveforms(
             self._metric_data._recording,
             self._metric_data._sorting,
             unit_ids=self._metric_data._unit_ids,
-            max_spikes_per_unit=max_spikes_per_unit_for_noise_overlap,
             **kwargs
         )
 
+        # set random seed
```

Comment:
if waveforms are already computed, then retrieving them is very quick. The problem is (and I think this is why you initially set `recompute_info=True`) that they could have been computed on less than `max_spikes_per_unit_for_noise_overlap`. In this case, I would do something like this:

```
n_waveforms_per_unit = [len(wf) for wf in waveforms]
n_spikes_per_unit = [len(self._metric_data._sorting.get_unit_spike_train(u)) for u in self._metric_data._unit_ids]

if np.all(n_waveforms_per_unit < max_spikes_per_unit_for_noise_overlap) and 
    np.any(n_spikes_per_unit > max_spikes_per_unit_for_noise_overlap):
    # in this case it means that waveforms have been computed on 
    # less spikes than max_spikes_per_unit_for_noise_overlap --> recompute
    kwargs['recompute_info'] = True
    waveforms = st.postprocessing.get_unit_waveforms(
            self._metric_data._recording,
            self._metric_data._sorting,
            unit_ids=self._metric_data._unit_ids,
            max_spikes_per_unit=max_spikes_per_unit_for_noise_overlap,
            **kwargs
        )
elif np.all(n_waveforms_per_unit > max_spikes_per_unit_for_noise_overlap):
    # waveforms computed on more spikes than needed --> sample
    for i_w, wfs in enumerate(waveforms):
        if len(wfs) > max_spikes_per_unit_for_noise_overlap:
                selecte_idxs = np.random.choice(times, size=max_spikes_per_unit_for_noise_overlap)
                waveforms[i] = wfs[selecte_idxs]
```

This makes sure that you always have `max_spikes_per_unit_for_noise_overlap` spikes without recomputing waveforms (if not needed).

Reply:
Yes that's precisely why I added `recompute_info=True`. Added your suggestion.